src/backend/distributed/metadata/pg_get_object_address_12_13_14.c

```diff
@@ -0,0 +1,536 @@
+
+	/* CITUS CODE END */
+
+	/* We don't need the relcache entry, thank you very much */
+	if (relation)
+	{
+		relation_close(relation, AccessShareLock);
+	}
+
+	/* CITUS CODE BEGIN */
+	return addr;
+
+	/* CITUS CODE END */
+}
+
+
+/*
+ * ErrorIfCurrentUserCanNotDistributeObject checks whether current user can
+ * distribute object, if not errors out.
+ */
+static void
+ErrorIfCurrentUserCanNotDistributeObject(ObjectType type, ObjectAddress *addr,
+										 Node *node, Relation *relation)
+{
+	Oid userId = GetUserId();
+	AclMode aclMaskResult = 0;
+	bool passAclCheck = false;
+	Oid idToCheck = InvalidOid;
+
+	/* Since we don't handle sequences like object, add it separately */
+	if (!(SupportedDependencyByCitus(addr) || type == OBJECT_SEQUENCE))
+	{
+		ereport(ERROR, (errmsg("Object type %d can not be distributed by Citus", type)));
+	}
+
+	switch (type)
+	{
+		case OBJECT_ACCESS_METHOD:
+		{
+			if (IsObjectAddressOwnedByExtension(addr, NULL))
+			{
+				ereport(ERROR, (errmsg("Current user does not have required "
+									   "access privileges on access method %d with type %d",
+									   addr->objectId, type)));
+			}
+			passAclCheck = true;
+			break;
+		}
+
+		case OBJECT_SCHEMA:
+		{
+			idToCheck = addr->objectId;
+			aclMaskResult = pg_namespace_aclmask(idToCheck, userId, ACL_USAGE,
+												 ACLMASK_ANY);
+			break;
+		}
+
+		case OBJECT_FUNCTION:
+		case OBJECT_PROCEDURE:
+		case OBJECT_AGGREGATE:
+		{
+			idToCheck = addr->objectId;
+			aclMaskResult = pg_proc_aclmask(idToCheck, userId, ACL_EXECUTE,
+											ACLMASK_ANY);
+			break;
+		}
+
+		case OBJECT_DATABASE:
+		{
+			idToCheck = addr->objectId;
+			aclMaskResult = pg_database_aclmask(idToCheck, userId, ACL_CONNECT,
+												ACLMASK_ANY);
+			break;
+		}
+
+		case OBJECT_ROLE:
+		{
+			/* Support only extension owner role with community */
+			if (!(addr->objectId == CitusExtensionOwner()))
+			{
+				ereport(ERROR, (errmsg("Current user does not have required "
+									   "access privileges on role %d with type %d",
+									   addr->objectId, type)));
+			}
+			passAclCheck = true;
+			break;
+		}
+
+		case OBJECT_TYPE:
+		{
+			idToCheck = addr->objectId;
+			aclMaskResult = pg_type_aclmask(idToCheck, userId, ACL_USAGE,
+											ACLMASK_ANY);
+			break;
+		}
+
+		case OBJECT_SEQUENCE:
+		case OBJECT_TABLE:
+		{
+			check_object_ownership(userId, type, *addr, node, *relation);
+			passAclCheck = true;
+			break;
+		}
+
+		case OBJECT_EXTENSION:
+		{
+			if (userId != CitusExtensionOwner())
+			{
+				ereport(ERROR, (errmsg("Only owner of the extension can distribute it")));
+			}
+			passAclCheck = true;
+			break;
+		}
+
+		case OBJECT_COLLATION:
+		{
+			passAclCheck = true;
+			break;
+		}
+
+		default:
+		{
+			ereport(ERROR, (errmsg("%d object type is not supported within "
+								   "object propagation", type)));
+			break;
+		}
+	}
+
+	if (passAclCheck == false && aclMaskResult == ACL_NO_RIGHTS)
```

Comment:
minor: maybe `passAclCheck` -> `skipAclCheck` as thats a more common pattern in the code. And `if (!skipAclCheck && ..`)